src/providers/pkcs11/capability_discovery.rs

```diff
@@ -0,0 +1,123 @@
+#![allow(unused, trivial_numeric_casts)]
+use super::Provider;
+use crate::authenticators::ApplicationName;
+use crate::providers::pkcs11::to_response_status;
+use cryptoki::types::mechanism::Mechanism;
+use cryptoki::types::mechanism::MechanismInfo;
+use cryptoki::types::mechanism::MechanismType;
+use cryptoki::types::Ulong;
+use log::trace;
+use parsec_interface::operations::can_do_crypto;
+use parsec_interface::operations::can_do_crypto::CheckType;
+use parsec_interface::operations::psa_algorithm::Algorithm;
+use parsec_interface::operations::psa_key_attributes::Attributes;
+use parsec_interface::operations::psa_key_attributes::EccFamily;
+use parsec_interface::operations::psa_key_attributes::Type;
+use parsec_interface::requests::ResponseStatus::{InvalidEncoding, PsaErrorNotSupported};
+use parsec_interface::requests::Result;
+use std::convert::TryFrom;
+use std::ops::Deref;
+
+impl Provider {
+    pub(super) fn can_do_crypto_internal(
+        &self,
+        _app_name: ApplicationName,
+        op: can_do_crypto::Operation,
+    ) -> Result<can_do_crypto::Result> {
+        let attributes = op.attributes;
+        let check_type = op.check_type;
+        let supported_ecc_family_sizes = [192, 224, 256, 384, 512];
+        match attributes.key_type {
+            Type::EccKeyPair {
+                curve_family: EccFamily::SecpR1,
+            }
+            | Type::EccPublicKey {
+                curve_family: EccFamily::SecpR1,
+            } => {
+                if !(supported_ecc_family_sizes.contains(&attributes.bits)) {
+                    return Err(PsaErrorNotSupported);
+                }
+            }
+            Type::RsaKeyPair | Type::RsaPublicKey => (),
+            _ => return Err(PsaErrorNotSupported),
+        }
+        if attributes.policy.permitted_algorithms != Algorithm::None {
+            let supported_mechanisms: Vec<MechanismType> = self
+                .backend
+                .get_mechanism_list(self.slot_number)
+                .map_err(to_response_status)?;
+            let mechanism = Mechanism::try_from(attributes.policy.permitted_algorithms)
+                .map_err(to_response_status)?;
+            let mechanism_info: MechanismInfo = self
+                .backend
+                .get_mechanism_info(self.slot_number, mechanism.mechanism_type())
+                .map_err(to_response_status)?;
+            if !(supported_mechanisms.contains(&mechanism.mechanism_type())) {
+                return Err(PsaErrorNotSupported);
+            }
```

Comment:
That's pretty clean!
I am thinking, maybe we should put that (and the key size check below) inside the `use_check` function? And replace the content of the `if` with just a call to `use_check`?

In the `match` at the end of the function we can check for `CheckType::Use` that the algorithm is not `None`.